js/src/components/tree-select-control/index.js

```diff
@@ -46,13 +46,31 @@ import { ARROW_DOWN, ARROW_UP, ENTER, ESCAPE, ROOT_VALUE } from './constants';
  **/
 
 /**
- *
- * @typedef {Object} Option
+ * @typedef {Object} CommonOption
  * @property {string} value The value for the option
+ * @property {string} [key] Optional unique key for the Option. It will fallback to the value property if not defined
+ */
+
+/**
+ * @typedef {Object} BaseOption
  * @property {string} label The label for the option
  * @property {Option[]} [children] The children Option objects
- * @property {string} [key] Optional unique key for the Option. It will fallback to the value property if not defined
- **/
+ *
+ * @typedef {CommonOption & BaseOption} Option
+ */
```

Comment:
💅  I will use just `Option` instead of CommonOption and BaseOption. (Or maybe I missed the intent of this separation). 

@typedef {Object} Option
@property {string} label The label for the option
@property {string} value The value for the option
@property {string} [key] Optional unique key for the Option. It will fallback to the value property if not defined
@property {Option[]} [children] The children Option objects

💅  WDYT of renaming `InnerOption` as well as `BaseInnerOption` to `Leaf` and `BaseLeaf`

In my mind I see it this way:
Option is the "pure" item we send to the tree, the data itself, with label, value, and optional key and children.
Leaf is and enhanced Option with Tree props (BasLeaf).

Reply:
> I will use just `Option` instead of CommonOption and BaseOption. (Or maybe I missed the intent of this separation).

Option and InnerOption both have `label` and `children` properties, but the type and description of the properties are different. JSDoc does not override the same attributes when using `@typedef {Option & BaseInnerOption} InnerOption`, but combines them.

For example, using this JSDoc: 

```js
/**
 * @typedef {Object} Option
 * @property {string} label The label for the option.
 * @property {string} value The value for the option.
 * @property {string} [key] Optional unique key for the Option. It will fallback to the value property if not defined.
 * @property {Option[]} [children] The children Option objects.
 */

/**
 * @typedef {Object} BaseInnerOption
 * @property {string|JSX.Element} label The label string or label with highlighted react element for the option.
 * @property {InnerOption[]|undefined} children The children options. The options are filtered if in searching.
 * (omitted)
 * @typedef {Option & BaseInnerOption} InnerOption
 */
```

There will be a combined type and description:

![2022-05-25 12 15 46](https://user-images.githubusercontent.com/17420811/170179900-3022b31f-2a00-450e-ba3f-d8e3f9e2c9a9.png)

Furthermore, the nested reference produces inconsistent results.

![2022-05-25 12 19 40](https://user-images.githubusercontent.com/17420811/170179936-1238fa78-4a9b-48a9-b694-a3414b201dc1.png)
![2022-05-25 12 21 10](https://user-images.githubusercontent.com/17420811/170179947-bb4b5b8e-e9ec-4699-8949-87e95c328f18.png)

---

> WDYT of renaming `InnerOption` as well as `BaseInnerOption` to `Leaf` and `BaseLeaf`

In a tree data structure, the terminology of *leaf* node refers to the node with no children. But an `InnerOption` could be a parent node, both parent and child node, or a leaf node. Therefore, I don't think `Leaf` is a proper name.

Reply:
> In a tree data structure, the terminology of leaf node refers to the node with no children. But an InnerOption could be a parent node, both parent and child node, or a leaf node. Therefore, I don't think Leaf is a proper name.

👌  Right. Maybe `node` then. 

> There will be a combined type and description:

Interesting. In my IDE it does the intersection properly.

<img width="727" alt="Screenshot 2022-05-25 at 12 31 34" src="https://user-images.githubusercontent.com/5908855/170218293-6442b841-7daa-4d40-a3f4-b325ba376cf2.png">

My worries are that Inner, Base, Common prefixes could be kinda confusing but this is a non blocking issue.

Reply:
> Maybe `node` then.

`Node` is a more general and abstract naming. When viewing the code like `@param {Node} option`, I don't think it's clearer, and it also seems more irrelevant to the parameter name.

> My worries are that Inner, Base, Common prefixes could be kinda confusing

Sorry, I'm not quite sure how they cause confusion in the scope of these type definitions since the Base and Common prefixes are only used for combining the Option and InnerOption. But I'd like to know if you have other naming ideas to replace them to reduce your confusion.

Reply:
All the names I have in mind rename Option to Node for defining the items of the tree and differentiate them from the Options provided by the user.

Option = { OptionValue & OptionData } 
Node = { OptionValue & BaseNode } 

`<Options options="filteredOptions"> to be renamed to <Nodes nodes="filteredNodes">`

But that requires more refactor of the Options component that was named wrongly from scratch by me. (notice in the Options.js component we refer to Nodes in ArrowLeft/ArrowRight docs) 

But this is pretty much a subjective and minor issue (unsure even if it worth it all the refactor in renaming).

Reply:
I think the overall naming adjustments could be a separate PR. Together with, for example, the `value` prop could be internally renamed to `values` or `selectedValues`, and the `nodesExpanded` to `expandedValues`, etc.